src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -42,6 +43,16 @@ internal sealed partial class SolutionState
     public SolutionOptionSet Options { get; }
     public IReadOnlyList<AnalyzerReference> AnalyzerReferences { get; }
 
+    /// <summary>
+    /// Fallback analyzer config options by language. The set of languages does not need to match the set of langauges of projects included in the surrent solution snapshot.
+    /// </summary>
+    public ImmutableDictionary<string, StructuredAnalyzerConfigOptions> FallbackAnalyzerOptions { get; } = ImmutableDictionary<string, StructuredAnalyzerConfigOptions>.Empty;
+
+    /// <summary>
+    /// Number of projects in the solution of the given language.
+    /// </summary>
+    internal ImmutableDictionary<string, int> ProjectCountByLanguage { get; } = ImmutableDictionary<string, int>.Empty;
```

Comment:
doc why we need this?

Reply:
personally, if we truly need this, i would prefer it be computed.  we've found that keeping computed state perfectly in sync with the solution data is *hard* (we get it wrong all the time).

Reply:
We subscribe to workspace update event so that we can include (remove) options for an added/removed language. If we calculate it on every solution/project-level update event we can end up with N^2 alg, e.g. when a solution is being loaded. Maintaining this map is rather trivial since we only need to update it when project state is added/removed.

Reply:
note: my storng preference here (based directly on the bugs we've gotten in these excat codepaths) is: 

Make this a required parameter of the forking/branching code.  And tehn explicitly pass in the existing value in all the other codepaths, with a comment saying "project structure isn't changing, so we do not have to update this".  

I will accept pushback here.  But we've gotten the fork/branch logic wrong a *ton* of times with default arguments.  So i'm wary.  

--

I do accept the N^2 arg for not computing on the fly.

Reply:
Sounds good. I can make it required param.

Reply:
Will make this change in a follow up PR.